datastore/filesystem/filesystem.py

```diff
@@ -105,105 +177,137 @@ def relative_object_path(self, key):
+		"""Stores or replaces the data named by `key` with `value`
+		
+		Arguments
+		---------
+		key
+			Key naming the binary data slot to store at
+		value
+			Some stream yielding the data to store
+		
+		Raises
+		------
+		RuntimeError
+			The given ``key`` names a subtree, not a value OR the contains a
+			value item as part of the key path
 		"""
-		path = self.object_path(key)
-		if os.path.exists(path):
-			os.remove(path)
-
-		# TODO: delete dirs if empty?
-
-	def query(self, query):
-		"""Returns an iterable of objects matching criteria expressed in `query`
-		FSDatastore.query queries all the `.obj` files within the directory
-		specified by the query.key.
-
-		Args:
-		  query: Query object describing the objects to return.
-
-		Returns:
-		  Cursor with all objects matching criteria
+		path = trio.Path(self.object_path(key))
+		
+		# Ensure containing directory exists
+		parent = path.parent
+		try:
+			await parent.mkdir(parents=True, exist_ok=True)
+		except FileExistsError as exc:
+			# Should hopefully only happen if `object_extension` is `""`
+			raise RuntimeError(f"Key '{key}' requires containing directory "
+			                   f"'{parent}' to not be a value") from exc
+		
+		try:
+			async with await trio.open_file(path, "wb") as file:
+				chunk = await value.receive_some(DEFAULT_BUFFER_SIZE)
+				while chunk:
+					await file.write(chunk)
+					
+					chunk = await value.receive_some(DEFAULT_BUFFER_SIZE)
+		except IsADirectoryError as exc:
+			# Should only happen if `object_extension` is `""`
+			raise RuntimeError(f"Key '{key}' names a subtree, not a value") from exc
+	
+
+	async def delete(self, key: datastore.Key):
+		"""Removes the data named by `key`
+		
+		Arguments
+		---------
+		key
+			Key naming the binary data slot to remove
+		
+		Raises
+		------
+		KeyError
+			The given object was not present in this datastore
 		"""
-		path = self.path(query.key)
-
-		if os.path.exists(path):
-			filenames = os.listdir(path)
-			filenames = list(set(filenames) - set(self.ignore_list))
-			filenames = map(lambda f: os.path.join(path, f), filenames)
-			iterable = FileSystemDatastore._read_object_gen(filenames)
-		else:
-			iterable = list()
-
-		return query(iterable)  # must apply filters, etc naively.
-
-	def contains(self, key):
-		"""Returns whether the object named by `key` exists.
-		Optimized to only check whether the file object exists.
-
-		Args:
-		  key: Key naming the object to check.
-
-		Returns:
-		  boolean whether the object exists
-		"""
-		path = self.object_path(key)
-		return os.path.exists(path) and os.path.isfile(path)
+		path = trio.Path(self.object_path(key))
+		
+		try:
+			await path.unlink()
+		except FileNotFoundError as exc:
+			raise KeyError(key) from exc
+		
+		# Try to remove parent directories if they are empty
+		try:
+			parent = path.parent
+			while str(parent).startswith(self.root_path + os.path.sep) \
```

Comment:
I didn't understand the logic here, so no comments

Reply:
It tries to `rmdir` parent directories as long as

  1. that directory is a subdirectory of the `root_path`
     * that is: the directory path starts with the string `<…/some/root/path>/`
  2. and the directory's parent is not the directory itself again
     * this guards against the case of the root directory being the current directory (`.`) whose `.parent` property is also `.` (I'm am not sure anymore if this isn't subsumed the first rule however as `".".startswith("./")` is `False` either way…?

I probably should a comment to the code about this, shouldn't I?